### PR TITLE
Factory contracts and sync engine upgrades

### PR DESCRIPTION
**Pull request by 0xOlias**
This PR introduces factory contract support and makes some improvements to the sync cache design.

### WARNING
- Existing Ponder apps that upgrade to this version will need to re-sync their event store from scratch. Large apps that have a long sync time should consider testing the re-sync on a new branch/railway project before upgrading on main.

### How to test
1. To create a new project on the prerelease version -> `pnpm create ponder@next`
2. To upgrade an existing app -> `pnpm upgrade @ponder/core@next` (be warned, this will bust your local sync cache)
3. Open up the [new factory contracts documentation](https://ponder-docs-git-typedarray-sync-cache-upgrades-cantrip.vercel.app/advanced/factory-contracts), and try to sync/index your favorite factory contract!

### Fixes
- #356 
- #343 
- #334

### CODE REVIEW

**Comment by shrugs** [comment]
i imagine you've thought of this but i think it's be very useful for the handlers of a child contract to receive a contract in the context that's bound to the address, rather than accessing `event.log.address`

it may also be useful to give handlers some sort of `setup` handler for when a new child contract is added, for example to ensure that there's an appropriate entity representing that new contract in the database

```ts
// LSSVMPair.ts
ponder.on("new", async ({ event, context }) => {
  const { Pair, TokenContract } = context.entities;
  const { LSSVMPair } = context.contracts;

  const [owner, nft, fee, spotPrice, delta] = await Promise.all([
    LSSVMPair.read.owner(),
    LSSVMPair.read.nft(),
    LSSVMPair.read.fee(),
    LSSVMPair.read.spotPrice(),
    LSSVMPair.read.delta(),
  ]);

  await Promise.all([
    Pair.upsert({
      id: addressId(LSSVMPair.address),
      create: {
        fee,
        spotPrice,
        delta,
        nft: addressId(nft),
        owner: addressId(owner),
      },
      update: {},
    }),
    TokenContract.upsert({
      id: addressId(nft),
      create: { },
      update: {},
    }),
  ]);
});
```

**Comment by 0xOlias** [comment]
> it may also be useful to give handlers some sort of `setup` handler for when a new child contract is added, for example to ensure that there's an appropriate entity representing that new contract in the database

Just seeing this. Awesome idea. I wonder what the event should be called? Your idea seemed to suggest that we would infer the contract name from the file name (I think) which is not currently a pattern. The only "special" event name we have currently is "setup" - all others are `${sourceName}:${eventName}`. 

Rambling a bit now, but soon enough we'll also want to add transaction call events, which could look like `"UniswapPool.swap"` (using `.` as the separator instead of `:`).

In the case of this "child contract created" event, maybe we could use `_` as the separator (`"UniswapPool_created"`. Don't like that very much though.

**Comment by shrugs** [comment]
ah, didn't mean to imply file name detection, i just forgot to include the appropriate information in the event name. perhaps "::" as like a "class level" delimiter could work, but perhaps too easily confused with ":" for events. @ or | also exist but i'm less enthused by then

perhaps "setup:ContractName"? though setup maybe not be the right word for this

File: packages/core/src/utils/lowercase.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Transforms the input string to lower case.
+ */
+export function toLowerCase<T extends string>(value: T) {
+  return value.toLowerCase() as T;
```
Comment (kyscott18) [inline packages/core/src/utils/lowercase.ts:5]: There's a cool generic in typescript called Lowercase that you could use here. 

```ts
return value.toLowerCase() as Lowercase<T>;
```

Comment (0xOlias) [reply packages/core/src/utils/lowercase.ts]: Just pushed a commit that uses `Lowercase<T>` and updates those spots to use the util - thanks!

**Comment by shrugs** [comment]
oh and re: setup event, this could just be typed as a handler of the factory event itself, so perhaps to support this pattern one adds the factory itself as a contract, listens to the correct event, and uses that as the setup? this requires that the factory's handler runs before any handlers in the contract, which i imagine is already the case. i'll test it out

**Comment by shrugs** [comment]
bug report on `0.0.93-next.1`: multiple factory configs using the same event but with different parameters doesn't seem to work in my case— only the first contract definition has its handlers invoked. the llama example works for me out of the box

```ts
const event = parseAbiItem("event NewNftaddress pair, address nft)");

export const config: Config = {
  networks: [network],
  contracts: [
    {
      network: network.name,
      name: "LSSVMPair",
      abi: "./abis/LSSVMPair.json",
      startBlock: factoryConfig.startBlock,
      factory: {
        address: factoryConfig.address,
        event,
        parameter: "pair",
      },
    },
    {
      network: network.name,
      name: "NFT",
      abi: "./abis/NFT.json",
      startBlock: factoryConfig.startBlock,
      factory: {
        address: factoryConfig.address,
        event,
        parameter: "nft",
      },
    },
  ],
};

```

if i change the order of the contracts, it's always the first one that has its handlers invoked (i'm destroyng the ponder cache between runs)

**Review by 0xOlias — COMMENTED** [review]
LGTM
obama-giving-obama-award

**Pull request merged**